### PR TITLE
resolved: demote only the fallback servers in the global scope

### PR DESCRIPTION
**Pull request by rpigott**
This softens the behavior introduced in c83cd8da4916, and simultaneously expands the role of fallback servers so they are applied not only when there are no dns servers configured, but when all the configured dns servers are configured only for non-default-route links.

The intent is that the global FallbackDNS (instead of DNS) can now be used in conjunction with the per-link dns, providing a fallback behavior without introducing a scope overlap.

~~Fixes: c83cd8da4916 (resolved: demote the global unicast scope)~~ (Reverted)

---

Closes: #33835 (again)
Replaces: #34047

### CODE REVIEW

**Comment by rpigott** [comment]
@poettering How about this one?

**Comment by yuwata** [comment]
Nah, I do not like this. Fallback DNS servers should behave the same as DNS servers. This complicates the logic and easily triggers surprising results...

**Comment by rpigott** [comment]
@yuwata really? Based on https://github.com/systemd/systemd/issues/33835#issuecomment-2296101711 it actually sounds to me that this is (already) the intended behavior.

File: src/resolve/resolved-dns-server.c [hunk 1]
```diff
@@ -909,7 +909,7 @@ DnsServer *manager_get_dns_server(Manager *m) {
                  * servers */
 
                 HASHMAP_FOREACH(l, m->links)
-                        if (l->dns_servers) {
+                        if (l->dns_servers && l->default_route) {
```
Comment (yuwata) [inline src/resolve/resolved-dns-server.c:910]: So, when `l->default_route` is changed, `m->current_dns_server` should be unset, something like the way in `dns_server_new()`.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by yuwata** [comment]
> @yuwata really? Based on [#33835 (comment)](https://github.com/systemd/systemd/issues/33835#issuecomment-2296101711) it actually sounds to me that this is (already) the intended behavior.

Not sure, but after reconsidering this change, this may be one of the way we can take. I withdraw the previous opinion.

**Comment by yuwata** [comment]
BTW, if this is the way to go, I think it is better to first revert the previous commit, and then change to the new behavior.
Then, we can easily backport the change to stable (though, I am not sure we should backport such change).

**Comment by rpigott** [comment]
Thanks. I appreciate the review. I will wait for @poettering's opinion.

File: src/resolve/resolved-dns-server.c (hunk 1 above)
Comment (poettering) [inline src/resolve/resolved-dns-server.c:910]: this looks like a bug fix actually, i'd say, and deserves its own separate commit, i'd say

File: src/resolve/resolved-dns-scope.c [hunk 2]
```diff
@@ -779,7 +779,10 @@ DnsScopeMatch dns_scope_good_domain(
                         return DNS_SCOPE_NO;
 
                 /* Prefer suitable per-link scopes where possible */
-                return s->link ? DNS_SCOPE_MAYBE : DNS_SCOPE_LAST_RESORT;
+                if (!s->link && s->manager->current_dns_server->type == DNS_SERVER_FALLBACK)
```
Comment (poettering) [inline src/resolve/resolved-dns-scope.c]: maybe use:

```c
DnsServer *server = dns_scope_get_dns_server(s);
if (server && server->type == DNS_SERVER_FALLBACK) …
```

File: src/resolve/resolved-dns-scope.c [hunk 3]
```diff
@@ -779,7 +779,10 @@ DnsScopeMatch dns_scope_good_domain(
...
+                        return DNS_SCOPE_LAST_RESORT;
+
+                return DNS_SCOPE_MAYBE;
```
Comment (poettering) [inline src/resolve/resolved-dns-scope.c:785]: i think this change does make a ton of sense.

**Comment by poettering** [comment]
this approach definitely makes sense to me.

File: src/resolve/resolved-link-bus.c [hunk 4]
```diff
@@ -465,6 +465,10 @@ int bus_link_method_set_default_route(sd_bus_message *message, void *userdata, s
         if (l->default_route != b) {
                 l->default_route = b;
 
+                DnsServer *server = l->manager->current_dns_server;
+                if (server && server->type == DNS_SERVER_FALLBACK)
+                        manager_set_dns_server(l->manager, NULL);
```
Comment (yuwata) [inline src/resolve/resolved-link-bus.c]: The same thing needs to be done also in `link_update_default_route()`.

Maybe, better to introduce a tiny helper function `link_set_default_route(Link *l, bool b)` with comments why we need to unset the current DNS server if it is fallback one.

File: src/resolve/resolved-dns-scope.c [hunk 5]
```diff
@@ -778,6 +778,11 @@ DnsScopeMatch dns_scope_good_domain(
                 if (!dns_scope_is_default_route(s))
                         return DNS_SCOPE_NO;
 
+                /* Prefer suitable per-link scopes where possible */
+                DnsServer *server = dns_scope_get_dns_server(s);
+                if (server && server->type == DNS_SERVER_FALLBACK)
```
Comment (yuwata) [inline src/resolve/resolved-dns-scope.c]: maybe, better to introduce a tiny inline helper function `dns_server_is_fallback()`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/resolve/resolved-link-bus.c (hunk 4 above)
Comment (poettering) [inline src/resolve/resolved-link-bus.c]: i'd prefer to avoid the "reachover" from the Link code here to the Manager code. maybe add manager_reconsider_current_server() or so which we just call here?

also comments would be good explaining what's going on here, because not obvious

Comment (yuwata) [reply src/resolve/resolved-link-bus.c]: > i'd prefer to avoid the "reachover" from the Link code here to the Manager code. maybe add manager_reconsider_current_server() or so which we just call here?

Sounds nice, then we can use it also in `dns_server_new()`.

Comment (rpigott) [reply src/resolve/resolved-link-bus.c]: `manager_set_dns_server(m, NULL)` basically already is `manager_reconsider_current_server`. If you don't like the raw member access into manager, I think stuffing this logic in a `link_set_default_route` as suggested by @yuwata will handle that.

File: src/resolve/resolved-dns-scope.c [hunk 6]
```diff
@@ -778,6 +778,11 @@ DnsScopeMatch dns_scope_good_domain(
                 if (!dns_scope_is_default_route(s))
                         return DNS_SCOPE_NO;
 
+                /* Prefer suitable per-link scopes where possible */
+                DnsServer *server = dns_scope_get_dns_server(s);
+                if (dns_server_is_fallback(server))
```
Comment (yuwata) [inline src/resolve/resolved-dns-scope.c]: Not a matter, but `dns_server_is_fallback(dns_scope_get_dns_server(s))` is enough.

**Review by yuwata — APPROVED** [review]
LGTM.

**Comment by yuwata** [comment]
@poettering Please take a final look.

**Comment by poettering** [comment]
lgtm

**Comment by Winterhuman** [comment]
Does https://systemd.io/RESOLVED-VPNS need updating to reflect any changes in behaviour introduced by this PR, or is it identical to before? Regardless, it might be nice to include an extra section on there to explain how this fallback logic works in more detail (I don't understand the logic enough to send a PR myself unfortunately)

**Comment by rpigott** [comment]
@Winterhuman that content looks accurate to me. I don't think this change has any bearing on VPNs.

This fallback is only needed when we don't have any other suitable DNS servers to use — the user did not specify a desired server, the network did not inform us of a local dns via dhcp or any other protocol, the query name does not match the search or route domains of any link, and any vpn/other interfaces present that do have DNS configured are excluded from general use by the default-route=no flag.
